[Site Isolation] Add basic support for webarchive generation

szewai opened the pull request:
#### 29d6acce4d23b6012258b0587f9e289a60b75b21
<pre>
[Site Isolation] Add basic support for webarchive generation
<a href="https://bugs.webkit.org/show_bug.cgi?id=293087">https://bugs.webkit.org/show_bug.cgi?id=293087</a>
<a href="https://rdar.apple.com/151424329">rdar://151424329</a>

Reviewed by Ryosuke Niwa.

In existing implementation, when creating webarchive for a web page, the web process that hosts the web page will
collect data for all frames and put the data in a single webarchive. With site isolation, this will not work because
frames can be put in different web processes and frame data needs to be fetched from different processes. To support
webarchive genereation with site isolation, this patch makes UI process collect frame webarchives from all web processes
of the page, and combined them into one webarchive based on subframe identifiers in the webarchives.

* Source/WebCore/loader/archive/Archive.h:
(WebCore::Archive::mainResource const):
(WebCore::Archive::mainResource): Deleted.
* Source/WebCore/loader/archive/ArchiveResource.cpp:
(WebCore::ArchiveResource::create):
(WebCore::ArchiveResource::createWithData):
* Source/WebCore/loader/archive/ArchiveResource.h:
(WebCore::ArchiveResource::createWithData):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::LegacyWebArchive):
(WebCore::LegacyWebArchive::create):
(WebCore::LegacyWebArchive::createFromSelection):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView createWebArchiveDataWithCompletionHandler:]):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::getWebArchiveData):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::getWebArchive):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getWebArchives):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, CreateWebArchive)):

Canonical link: <a href="https://commits.webkit.org/295026@main">https://commits.webkit.org/295026@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5bed5487a5fe7c35015c5450aae9f19aae6f461c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103821 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23523 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13853 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109014 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54473 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105861 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23876 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32069 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78881 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106827 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18541 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93670 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59207 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18343 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11719 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53849 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88103 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11778 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111401 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30977 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22845 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87885 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31339 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89870 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87532 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22291 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32424 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10162 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/25341 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; compiling") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30905 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36217 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/30699 "Build was cancelled. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Running apply-patch; Checked out pull request; compiling") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34034 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32260 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.h:
```diff
@@ -43,18 +44,26 @@ struct SimpleRange;
-    WEBCORE_EXPORT static RefPtr<LegacyWebArchive> create(LocalFrame&);
-    WEBCORE_EXPORT static RefPtr<LegacyWebArchive> createFromSelection(LocalFrame*, bool saveScriptsFromMemoryCache = true);
-    WEBCORE_EXPORT static RefPtr<LegacyWebArchive> create(const SimpleRange&, bool saveScriptsFromMemoryCache = true);
+    enum class ShouldSaveSubframeArchives : bool { No, Yes };
```
`ShouldSaveSubframeArchives` is a pretty wordy / mouthful. How about `ShouldArchiveSubframes`?

rniwa commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1552,6 +1554,73 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+            }
+            return archive;
+        }
+        ~WebArchvieCallbackAggregator()
```
Nit: need a blank line before this.

rniwa commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1552,6 +1554,73 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+
+        void addResult(HashMap<WebCore::FrameIdentifier, Ref<WebCore::LegacyWebArchive>>&& frameArchives)
+        {
+            for (auto frameIdentifier : copyToVector(frameArchives.keys())) {
```
Do we really need to make a copy of keys()? It seems like we can just iterate over them??

rniwa commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1552,6 +1554,73 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+        void addResult(HashMap<WebCore::FrameIdentifier, Ref<WebCore::LegacyWebArchive>>&& frameArchives)
+        {
+            for (auto frameIdentifier : copyToVector(frameArchives.keys())) {
+                if (RefPtr frameArchive = frameArchives.take(frameIdentifier))
```
It seems rather unnecessary to use take here? Are you trying to avoid a ref-churn?

szewai replied:
Yes; will change to just iterate directly on map

rniwa commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1552,6 +1554,73 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
+        RefPtr<WebCore::LegacyWebArchive> completeFrameArchive(FrameIdentifier identifier)
+        {
+            RefPtr archive = m_frameArchives.take(identifier);
+            if (archive) {
```
Why not add an early exit for when `archive` is `nullptr` instead?

rniwa commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -314,6 +314,18 @@ Ref<LegacyWebArchive> LegacyWebArchive::create()
+Ref<LegacyWebArchive> LegacyWebArchive::create(Ref<ArchiveResource>&& mainResource, Vector<Ref<ArchiveResource>>&& subresources, Vector<FrameIdentifier>&& subframeIdentifiers)
+{
+    auto archive = create();
+    archive->setSubframeIdentifiers(WTFMove(subframeIdentifiers));
```
Why not add a new `create` function which takes a subframe identifiers?

rniwa commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
```diff
@@ -529,18 +541,32 @@ RefPtr<LegacyWebArchive> LegacyWebArchive::create(LocalFrame& frame)
-        if (!localChild)
+    Vector<FrameIdentifier> subframeIdentifiers;
+    for (unsigned index = 0; index < frame.tree().childCount(); ++index) {
+        RefPtr child = frame.tree().child(index);
```
This is an awfully inefficient O(n^2) way of iterating over all child frames.
We should iterate over `nextSibling()` instead.

rniwa commented on Source/WebCore/loader/archive/cf/LegacyWebArchive.h:
```diff
@@ -43,18 +44,26 @@ struct SimpleRange;
     WEBCORE_EXPORT RetainPtr<CFDataRef> rawDataRepresentation();
 
+    Ref<ArchiveResource> protectedMainResource() const { return *mainResource(); }
+    void setSubframeIdentifiers(Vector<FrameIdentifier>&& identifiers) { m_subframeIdentifiers = WTFMove(identifiers); }
```
Given this function is only ever called right after `create` function, I think it would be better to add a new create function which takes the vector as an input.

szewai committed:
[Site Isolation] Add basic support for webarchive generation
https://bugs.webkit.org/show_bug.cgi?id=293087
rdar://151424329

Reviewed by Ryosuke Niwa.

In existing implementation, when creating webarchive for a web page, the web process that hosts the web page will
collect data for all frames and put the data in a single webarchive. With site isolation, this will not work because
frames can be put in different web processes and frame data needs to be fetched from different processes. To support
webarchive genereation with site isolation, this patch makes UI process collect frame webarchives from all web processes
of the page, and combined them into one webarchive based on subframe identifiers in the webarchives.

* Source/WebCore/loader/archive/Archive.h:
(WebCore::Archive::mainResource const):
(WebCore::Archive::mainResource): Deleted.
* Source/WebCore/loader/archive/ArchiveResource.cpp:
(WebCore::ArchiveResource::create):
(WebCore::ArchiveResource::createWithData):
* Source/WebCore/loader/archive/ArchiveResource.h:
(WebCore::ArchiveResource::createWithData):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::LegacyWebArchive):
(WebCore::LegacyWebArchive::create):
(WebCore::LegacyWebArchive::createFromSelection):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView createWebArchiveDataWithCompletionHandler:]):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::getWebArchiveData):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::getWebArchive):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::getWebArchives):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, CreateWebArchive)):

Canonical link: https://commits.webkit.org/295026@main
